mkosi

```diff
@@ -2263,76 +2323,85 @@ def install_arch(args: CommandLineArguments, root: str, do_run_build_script: boo
-                        """
-                    )
-                )
-
     # Install the user-specified packages and kernel
     packages = set(args.packages)
-    if not do_run_build_script and args.bootable:
-        packages |= kernel_packages
+    if not do_run_build_script and args.bootable and not any(["linux" in package for package in args.packages]):
+        # No user-specified kernel
+        packages.add("linux")
 
     if do_run_build_script:
         packages.update(args.build_packages)
 
-    def run_pacman(args: List[str], **kwargs: Any) -> CompletedProcess:
-        cmdline = [
-            "pacman",
-            "--noconfirm",
-            "--color", "never",
-            "--config", pacman_conf,
-        ]
+    if not do_run_build_script and args.bootable:
+        hooks_dir = os.path.join(root, "etc/pacman.d/hooks")
+        os.makedirs(hooks_dir, 0o755, exist_ok=True)
 
-        try:
-            return run(cmdline + args, **kwargs, check=True)
-        finally:
-            stop_gpg_agent()
+        # Disable depmod pacman hook as depmod is handled by kernel-install as well.
+        os.symlink("/dev/null", os.path.join(hooks_dir, "60-depmod.hook"))
 
-    # Remove already installed packages
-    c = run_pacman(['-Qq'], stdout=PIPE, universal_newlines=True)
-    packages.difference_update(c.stdout.split())
+        kernel_install_hook = os.path.join(hooks_dir, "kernel-install.hook")
+        with open(kernel_install_hook, 'w') as f:
+            f.write(
+                dedent(
+                    f"""
+                    [Trigger]
+                    Operation = Install
+                    Operation = Upgrade
+                    Type = File
+                    Target = usr/lib/modules/*/vmlinuz
+
+                    [Action]
+                    Description = Adding kernel and initramfs images to /boot...
+                    When = PostTransaction
+                    Exec = /bin/bash -c 'while read f; do kernel-install add $(basename $(dirname $f)) $f; done'
+                    Depends = systemd
+                    Depends = dracut
+                    Depends = bash
+                    Depends = coreutils
+                    NeedsTargets
+                    """
+                )
+            )
 
-    if packages:
-        # Disable mkinitcpio hook because we manually install the kernel image
-        # to /boot using kernel-install in install_boot_loader_arch.
-        hooks = os.path.join(root, "etc/pacman.d/hooks")
-        os.makedirs(hooks, 0o755, exist_ok=True)
+        kernel_remove_hook = os.path.join(hooks_dir, "kernel-remove.hook")
+        with open(kernel_remove_hook, 'w') as f:
+            f.write(
+                dedent(
+                    f"""
+                    [Trigger]
+                    Operation = Upgrade
+                    Operation = Remove
+                    Type = File
+                    Target = usr/lib/modules/*/vmlinuz
```

Comment:
Just to start a list on which packages we should rebuild the initrd on: linux-*, systemd, dracut, intel-ucode, amd-ucode, lvm2, device-mapper, cryptsetup. I probably forget all the RAID packages I never use, but since mkosi doesn't target that, we can probably do without them (which would also be true for LVM).

Reply:
So we'd make a separate pacman hook and in the hook check whether there's paths that don't end with vmlinuz. If there are, rebuild everything, if not, loop over the given paths and only run kernel-install on those?

Reply:
I don't think a separate hook, but instead of this hook calling kernel-install calling that script instead. Apart from that, yes, loop over the input to see whether it's a kernel package or something else and on that basis rebuild the kernel image.

Reply:
Ah, and of course this should be in the upgrade hook instead of this one.

Reply:
This should of course be in the add Installation hook instead of the Remova hook.